src/model/buffer.rs

```diff
@@ -72,267 +28,381 @@ pub struct Buffer {
-            cursor.node_offset = new_node.offset();
-            cursor.line_idx = new_node.line_offsets_len() - 1;
-            cursor.line_offset = cursor.node_offset - new_node.last_line_offset();
-            self.node_list.push(new_node);
-        } else if cursor.node_offset == 0 {
+            self.cursor.node_offset = new_node.offset();
+            self.cursor.line_idx = new_node.line_offsets_len() - 1;
+            self.cursor.line_offset = line_idx_update(self.cursor.line_offset);
+            self.cursor.original_line_offset = self.cursor.line_offset;
+
+            self.node_list.insert_curr(new_node);
+        } else if self.cursor.node_offset == 0 {
             // cursor should be at the front of a node
-            assert_eq!(cursor.node_offset, 0);
-            assert_eq!(cursor.line_idx, 0);
-            assert_eq!(cursor.line_offset, 0);
+            assert_eq!(self.cursor.node_offset, 0);
+            assert_eq!(self.cursor.line_idx, 0);
             // just insert the new node before current node
             //	which is where the cursor is
-            self.node_list.insert(cursor.node_idx, new_node);
-            cursor.node_idx += 1;
+            self.cursor.node_offset = new_node.offset();
+            self.cursor.line_idx = new_node.line_offsets_len() - 1;
+            self.cursor.line_offset = line_idx_update(self.cursor.line_offset);
+            self.cursor.original_line_offset = self.cursor.line_offset;
+            self.node_list.insert_prev(new_node);
         } else {
-            // split the node into two and insert it in between
+            // split the node into two and replace the current node with them
             let from = node_to_split.from();
             let index = node_to_split.index();
             let offset = node_to_split.offset();
             let line_offsets = &mut node_to_split.line_offsets();
 
             // split the line_offsets based on where the cursor is
-            let left_line_offsets = line_offsets.drain(..=cursor.line_idx).collect();
+            let cursor_node_offset = self.cursor.node_offset;
+            let left_line_offsets = line_offsets.drain(..=self.cursor.line_idx).collect();
             let mut right_line_offsets: Vec<usize> = line_offsets
                 .drain(..)
-                .map(|x| x - cursor.node_offset)
+                .map(|x| x - cursor_node_offset)
                 .collect();
             right_line_offsets.insert(0, 0);
 
             // construct left and right nodes
-            let left_node = BufferNode::new(from, index, cursor.node_offset, left_line_offsets);
+            let left_node =
+                BufferNode::new(from, index, self.cursor.node_offset, left_line_offsets);
             let right_node = BufferNode::new(
                 from,
-                index + cursor.node_offset,
-                offset - cursor.node_offset,
+                index + self.cursor.node_offset,
+                offset - self.cursor.node_offset,
```

Comment:
Minor point but, since you defined `let cursor_node_offset = self.cursor.node_offset` in line 157, you could just use `cursor_node_offset` instead of `self.cursor.node_offset` in these 3 lines, right?

Reply:
ah right, thanks!